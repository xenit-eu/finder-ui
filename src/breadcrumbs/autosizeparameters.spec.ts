import generateParameters, { GenerateParametersOutput_t } from "./autosizeparameters";

function toArray<T>(generator: Generator<T>): readonly T[] {
    const data: T[] = [];
    let value = generator.next();

    while (!value.done) {
        data.push(value.value);
        value = generator.next();
    }

    return data;
}

function sortByPriority(params: readonly GenerateParametersOutput_t[]): readonly GenerateParametersOutput_t[] {
    return params.slice().sort((a, b) => b.priority - a.priority);
}

describe("breadcrumbs/autosizeparameters", () => {
    it("generates one set of parameters when maxItems is set", () => {
        const data = toArray(
            generateParameters({
                size: 10,
                maxItems: 4,
            }),
        );

        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when no constraints are set", () => {
        const data = toArray(
            generateParameters({
                size: 10,
            }),
        );

        expect(data).toContainEqual({
            priority: 11,
            output: {
                maxItems: 10,
                itemsBeforeCollapse: 1,
                itemsAfterCollapse: 1,
            },
        });
        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having itemsBeforeCollapse fixed", () => {
        const data = toArray(
            generateParameters({
                size: 10,
                itemsBeforeCollapse: 1,
            }),
        );

        expect(data.map((d) => d.output.itemsBeforeCollapse)).not.toContain(2);

        expect(sortByPriority(data.slice(1)).map((d) => d.output.itemsAfterCollapse)).toEqual([9, 8, 7, 6, 5, 4, 3, 2, 1]);

        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having itemsAfterCollapse fixed", () => {
        const data = toArray(
            generateParameters({
                size: 10,
                itemsAfterCollapse: 2,
            }),
        );

        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(1);
        expect(data.map((d) => d.output.itemsAfterCollapse)).not.toContain(3);

        expect(sortByPriority(data.slice(1)).map((d) => d.output.itemsBeforeCollapse)).toEqual([8, 7, 6, 5, 4, 3, 2, 1]);

        expect(sortByPriority(data)).toMatchSnapshot();

    });

    it("generates parameters when having itemsBeforeCollapse set to a range", () => {
        const data = toArray(
            generateParameters({
                size: 10,
                itemsBeforeCollapse: {
                    min: 1,
                    max: 2,
                },
            }),
        );

        expect(data.slice(1).map((d) => d.output.itemsBeforeCollapse)).not.toContain(3);
        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having itemsAfterCollapse set to a range", () => {
        const data = toArray(
            generateParameters({
                size: 10,
                itemsAfterCollapse: {
                    min: 2,
                    max: 4,
                },
            }),
        );

        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(1);
        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(5);
        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having both itemsBeforeCollapse and itemsAfterCollapse set to a range", () => {
        const data = toArray(
            generateParameters({
                size: 10,
                itemsBeforeCollapse: {
                    min: 1,
                    max: 2,
                },
                itemsAfterCollapse: {
                    min: 1,
                    max: 2,
                },
            }),
        );
        expect(data.slice(1).map((d) => d.output.itemsBeforeCollapse)).toContain(2);
        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).toContain(2);
        expect(data.slice(1).map((d) => d.output.itemsBeforeCollapse)).not.toContain(3);
        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(3);
        expect(sortByPriority(data)).toMatchSnapshot();
    });
});
