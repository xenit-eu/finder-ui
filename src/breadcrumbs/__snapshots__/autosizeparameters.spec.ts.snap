// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breadcrumbs/autosizeparameters generates one set of parameters when maxItems is set 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 4,
    },
    "priority": 1,
  },
]
`;

exports[`breadcrumbs/autosizeparameters generates parameters when having both itemsBeforeCollapse and itemsAfterCollapse set to a range 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 4,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 2,
      "maxItems": 3,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 2,
    },
    "priority": 2,
  },
]
`;

exports[`breadcrumbs/autosizeparameters generates parameters when having itemsAfterCollapse fixed 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 8,
      "maxItems": 10,
    },
    "priority": 10,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 7,
      "maxItems": 9,
    },
    "priority": 9,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 6,
      "maxItems": 8,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 5,
      "maxItems": 7,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 4,
      "maxItems": 6,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 3,
      "maxItems": 5,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 4,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 3,
    },
    "priority": 3,
  },
]
`;

exports[`breadcrumbs/autosizeparameters generates parameters when having itemsAfterCollapse set to a range 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 6,
      "maxItems": 10,
    },
    "priority": 10,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 6,
      "maxItems": 9,
    },
    "priority": 9,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 5,
      "maxItems": 8,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 4,
      "maxItems": 7,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 3,
      "maxItems": 6,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 2,
      "maxItems": 5,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 4,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 3,
    },
    "priority": 3,
  },
]
`;

exports[`breadcrumbs/autosizeparameters generates parameters when having itemsBeforeCollapse fixed 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 9,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 10,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 8,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 9,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 7,
      "itemsBeforeCollapse": 1,
      "maxItems": 8,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 6,
      "itemsBeforeCollapse": 1,
      "maxItems": 7,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 5,
      "itemsBeforeCollapse": 1,
      "maxItems": 6,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 1,
      "maxItems": 5,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 1,
      "maxItems": 4,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 3,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 2,
    },
    "priority": 2,
  },
]
`;

exports[`breadcrumbs/autosizeparameters generates parameters when having itemsBeforeCollapse set to a range 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 8,
      "itemsBeforeCollapse": 2,
      "maxItems": 10,
    },
    "priority": 10,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 7,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 9,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 6,
      "itemsBeforeCollapse": 2,
      "maxItems": 8,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 5,
      "itemsBeforeCollapse": 2,
      "maxItems": 7,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 2,
      "maxItems": 6,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 2,
      "maxItems": 5,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 4,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 2,
      "maxItems": 3,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 2,
    },
    "priority": 2,
  },
]
`;

exports[`breadcrumbs/autosizeparameters generates parameters when no constraints are set 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 5,
      "itemsBeforeCollapse": 5,
      "maxItems": 10,
    },
    "priority": 10,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 5,
      "maxItems": 9,
    },
    "priority": 9,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 4,
      "maxItems": 8,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 4,
      "maxItems": 7,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 3,
      "maxItems": 6,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 3,
      "maxItems": 5,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 4,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 2,
      "maxItems": 3,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 2,
    },
    "priority": 2,
  },
]
`;
